### ISSUE TITLE
IRC logs, first feedback on Image dataformat

### ISSUE THREAD

**Issue by Kubuxu** [jonnor/ipld-image#1]
I had talk with @jonnor over #ipfs@freenode and here are the logs:

> 2016-06-03 17:36:22 @Kubuxu jonnor: where do you accept comments, I can give few right away.
> 2016-06-03 17:39:22 jonnor  Kubuxu, PR? or we can discuss right here and now
> 2016-06-03 17:39:40 @Kubuxu Ok, so here it goes:
> 2016-06-03 17:41:12 @Kubuxu Storageof links to roots of tile lists on the top level would be the best IMO, it allows for better deduplication (when for example metadata changes).
> 2016-06-03 17:42:35 @Kubuxu PNG would be good choice, but I would define in format that tiles HAVE TO be losslessly compressed using given alg (for convergence) 
> 2016-06-03 17:43:18 @Kubuxu In case of not squere images, there could be a "default" tile defined.
> 2016-06-03 17:43:19 jonnor  Kubuxu,  good point. Do you think one link per level, or one link to to the entire pyramid would make most sense? to make sense the whole pyramid would need to change at same time, since the data are derived from eachother
> 2016-06-03 17:43:59 @Kubuxu not rectangular **
> 2016-06-03 17:44:08 @Kubuxu That would be used if there is no tile was assigned to that spot.
> 2016-06-03 17:44:56 @Kubuxu re mipmaps links: one link to list of links that point to whole mipmaps
> 2016-06-03 17:44:58 jonnor  specifyable default tile makes makes sense, much more than trying to encode 'valid options'
> 2016-06-03 17:45:44 jonnor  Kubuxu, ok. Can't a lot of indirection become a potential perf issue?
> 2016-06-03 17:46:45 @Kubuxu bitswap is supposed to made smart enough to do something like that: you ask me about hash A, people who ask me about hash A return almost always asking about hash B and C, so I send you A B and C right away.
> 2016-06-03 17:47:13 jonnor  cool
> 2016-06-03 17:48:07 @Kubuxu I would define that tile always has to fit in one block (sub 256KB - ~2K).
> 2016-06-03 17:49:21 jonnor  regarding lossless/lossy. I agree that lossless makes sense for vast majority, and PNG will likely do great. The idea of allowing lossy would be for serving a JPEG/WebP directly to browsers
> 2016-06-03 17:49:41 jonnor  by concatinating a selected set of tiles (at a given mipmap level)
> 2016-06-03 17:51:14 @Kubuxu Problem with lossy compression is that it will always have problems: either you compress whole image and you will almost never be able to dedup tiles or you comress tiles then the compression is much worst/it looks even worst.
> 2016-06-03 17:52:20 @Kubuxu With lossless compression on tile level you have dedupable tiles, compression ratios might be a bit worst but tiles are reusable and look as they should.
> 2016-06-03 17:53:34 jonnor  yep. But lossy images can be so much smaller compared to lossless...
> 2016-06-03 17:54:38 @Kubuxu Then we should at least have flag that image is already lossy compressed, so we don't have the JPEG effect.
> 2016-06-03 17:54:57 @Kubuxu https://www.youtube.com/watch?v=QEzhxP-pdos
> 2016-06-03 17:55:05 jonnor  yes, it could be treated as an 'output only' type/variant
> 2016-06-03 17:55:32 jonnor  which should probably have a 'derivative of' link
> 2016-06-03 17:55:35 @Kubuxu hmm, and it could also like to the original, wow, that would be awesome
> 2016-06-03 17:55:37 jonnor  back to the lossless structure
> 2016-06-03 17:56:58 @Kubuxu Storage of all mipmaps (up to 1x1 or 1xN) requires only 33% on average more space.
> 2016-06-03 17:57:13 @Kubuxu So I would say it is really worth it.
> 2016-06-03 17:57:35 jonnor  aside: in imgflo, which is the image processing server we use for these things now, we also do effects/filtering/whatever processing. And this is described as a mapping between the input image, and algorithm, and a set of parameters
> 2016-06-03 17:58:08 jonnor  so then it would be possible to express also a processed image with backlinks to the original
> 2016-06-03 17:58:14 jonnor  in IPLD
> 2016-06-03 17:58:57 @Kubuxu Especially that you could load images in /8 screen resolution at first (as preview) which requires 1/64 bandwidth.
> 2016-06-03 18:04:39 jonnor  do you think a generic 'derivedFrom'  backlink would be useful
> 2016-06-03 18:05:24 @Kubuxu I think yes, if people were using it, it would be awesome.
> 2016-06-03 18:05:35 jonnor  with that one one will not neccesarily know how this thing was derived (what algorithm etc), but at least know that there is another previous version, and which one it was
> 2016-06-03 18:06:04 @Kubuxu Also format you proposed has no downsides to including higher resolutions with the image itself, which is awesome.
> 2016-06-03 18:06:51 @Kubuxu You might also think about allowing lower resolution alternatives, apart from the mipmaps.
> 2016-06-03 18:07:35 @Kubuxu Like imagine wallpaper, 4k version, 1080 version and 720 version. Our computer will download and use whichever first the best.
> 2016-06-03 18:08:10 @Kubuxu Problem is that it would require some changes who we view images (as you might want to say how big you want it before starting download).
> 2016-06-03 18:08:21 jonnor  that can be done by the client?
> 2016-06-03 18:09:43 @Kubuxu What I think: Browser knows how big is space for the image it will be downloading, if that information was passed along, you could download the best match.
> 2016-06-03 18:11:08 jonnor  by space, do you mean the space it will take on-screen (in that particular layout)?
> 2016-06-03 18:13:08 @Kubuxu yup
> 2016-06-03 18:13:48 jonnor  I don't think browsers will be IPFS aware anytime soon... unfortunately
> 2016-06-03 18:14:02 @Kubuxu But JS inside the browser can be.
> 2016-06-03 18:15:19 @Kubuxu Like imagine js code like that: There is local IPFS node running -> I have picture to load that is also available on IPFS -> I have on screen size of that picture -> Use local IPFS node, download needed size.
> 2016-06-03 18:15:23 jonnor  I think one should generate picture elements (and use media-query) with different URLs, each being an IPFS hash/path
> 2016-06-03 18:15:43 jonnor  and maybe add some JS to do even better than that, like progressive load
> 2016-06-03 18:17:39 jonnor  I'm _hoping_, based on chat today with jbenet, that we can use an IPLD path expression to say "construct an image file" from this ipfs object, at a certain mipmap level (and thus resolution)

**Comment by jonnor** [comment]
Great for transparency to have this in github. I believe I've addressed your specific feedback points in:
https://github.com/jonnor/ipld-image/commit/714621c0a41e41799135335112885b66a20e2284
https://github.com/jonnor/ipld-image/commit/5a51bd5946dad46384b0b5cc857ea246c8d97253
https://github.com/jonnor/ipld-image/commit/5a51bd5946dad46384b0b5cc857ea246c8d97253

Should we close this (and then open another issue for new logs)? Or do we keep this open and put relevant new discussions also here? Unsure of best practice

**Comment by Kubuxu** [comment]
I think we can close it. It will be there as reference for the future and in case we want to read it again.
